File: src/bitarray.cpp
#include "bitarray.h"
#include <algorithm>
#include <stdexcept>
#include <sstream>
#include <iomanip>
#include <functional>
#include <limits>

#ifdef _MSC_VER
#include <intrin.h>
inline int ctzll(unsigned __int64 value) {
    unsigned long index;
    return _BitScanForward64(&index, value) ? index : 64;
}
inline int popcntll(unsigned __int64 value) {
    return __popcnt64(value);
}
#else
#define ctzll(x) __builtin_ctzll(x)
#define popcntll(x) __builtin_popcountll(x)
#endif

// Constructor
BitArray::BitArray(size_t size) : bits(size), disposed(false) {
    if (size > std::numeric_limits<size_t>::max() - 63) {
        throw std::out_of_range("Requested size overflows capacity");
    }
    size_t words = size / 64 + (size % 64 != 0 ? 1 : 0);
    data.resize(words, 0); // Allocate 64-bit words for the given number of bits
}

// Destructor
BitArray::~BitArray() {
    dispose();
}

// Dispose the instance
void BitArray::dispose() {
    data.clear();
    bits = 0;
    disposed = true;
}

// Helper function to validate a position
void BitArray::validate(size_t pos) const {
    if (disposed) throw std::runtime_error("Storage is disposed");
    if (pos >= bits) throw std::out_of_range("Invalid position: " + std::to_string(pos));
}

// Helper function to calculate word index and bit offset
std::pair<size_t, size_t> BitArray::getPos(size_t pos) const {
    return {pos / 64, pos % 64};
}

// Helper function to clear unused padding bits in the last word
void BitArray::clearUnusedBits() {
    if (disposed || data.empty()) return;
    size_t lastWordBits = bits % 64;
    if (lastWordBits > 0) {
        data.back() &= ((static_cast<uint64_t>(1) << lastWordBits) - 1);
    }
}

// Set a bit at a specific position
void BitArray::set(size_t pos, bool value) {
    validate(pos);
    auto [index, offset] = getPos(pos);
    if (value) {
        data[index] |= (static_cast<uint64_t>(1) << offset); // Set the bit
    } else {
        data[index] &= ~(static_cast<uint64_t>(1) << offset); // Clear the bit
    }
}

// Get a bit at a specific position
bool BitArray::get(size_t pos) const {
    validate(pos);
    auto [index, offset] = getPos(pos);
    return (data[index] & (static_cast<uint64_t>(1) << offset)) != 0; // Check if the bit is set
}

// Resize the BitArray
void BitArray::resize(size_t newSize) {
    if (disposed) throw std::runtime_error("Storage is disposed");
    if (newSize > std::numeric_limits<size_t>::max() - 63) {
        throw std::out_of_range("Requested size overflows capacity");
    }
    size_t words = newSize / 64 + (newSize % 64 != 0 ? 1 : 0);
    std::vector<uint64_t> newData(words, 0);
    std::copy(data.begin(), data.begin() + std::min(data.size(), newData.size()), newData.begin());
    data = std::move(newData);
    bits = newSize;
    clearUnusedBits();
}

// Perform bitwise operations
void BitArray::bitwiseOp(const BitArray& other, std::function<uint64_t(uint64_t, uint64_t)> op) {
    if (bits != other.bits) throw std::invalid_argument("Size mismatch");
    for (size_t i = 0; i < data.size(); ++i) {
        data[i] = op(data[i], other.data[i]);
    }
}

// Specific bitwise operations
void BitArray::bitwiseAnd(const BitArray& other) {
    if (bits != other.bits) throw std::invalid_argument("Size mismatch");
    for (size_t i = 0; i < data.size(); ++i) {
        data[i] &= other.data[i];
    }
}

void BitArray::bitwiseOr(const BitArray& other) {
    if (bits != other.bits) throw std::invalid_argument("Size mismatch");
    for (size_t i = 0; i < data.size(); ++i) {
        data[i] |= other.data[i];
    }
}

void BitArray::bitwiseXor(const BitArray& other) {
    if (bits != other.bits) throw std::invalid_argument("Size mismatch");
    for (size_t i = 0; i < data.size(); ++i) {
        data[i] ^= other.data[i];
    }
}

void BitArray::bitwiseNot() {
    for (size_t i = 0; i < data.size(); ++i) {
        data[i] = ~data[i];
    }
    clearUnusedBits();
}

// Set multiple bits at once
void BitArray::setBatch(const std::vector<size_t>& positions, bool value) {
    if (disposed) throw std::runtime_error("Storage is disposed");
    for (auto pos : positions) {
        if (pos >= bits) throw std::out_of_range("Invalid position: " + std::to_string(pos));
    }
    if (value) {
        for (auto pos : positions) {
            data[pos / 64] |= (static_cast<uint64_t>(1) << (pos % 64));
        }
    } else {
        for (auto pos : positions) {
            data[pos / 64] &= ~(static_cast<uint64_t>(1) << (pos % 64));
        }
    }
}

// Get multiple bits at once
std::vector<bool> BitArray::getBatch(const std::vector<size_t>& positions) const {
    if (disposed) throw std::runtime_error("Storage is disposed");
    std::vector<bool> result;
    result.reserve(positions.size());
    for (auto pos : positions) {
        if (pos >= bits) throw std::out_of_range("Invalid position: " + std::to_string(pos));
        result.push_back((data[pos / 64] & (static_cast<uint64_t>(1) << (pos % 64))) != 0);
    }
    return result;
}

// Set a range of bits
void BitArray::setRange(size_t min, size_t max, bool value) {
    validate(min);
    validate(max);
    if (min > max) throw std::invalid_argument("min cannot be greater than max");

    auto [startWord, startOffset] = getPos(min);
    auto [endWord, endOffset] = getPos(max);

    if (startWord == endWord) {
        size_t k = endOffset - startOffset + 1;
        uint64_t mask = (k == 64) ? ~0ULL : ((1ULL << k) - 1);
        mask <<= startOffset;
        if (value) {
            data[startWord] |= mask;
        } else {
            data[startWord] &= ~mask;
        }
    } else {
        // 1. Modify partial start word
        uint64_t startMask = ~0ULL << startOffset;
        if (value) {
            data[startWord] |= startMask;
        } else {
            data[startWord] &= ~startMask;
        }

        // 2. Modify fully spanned intermediate words
        uint64_t fillVal = value ? ~0ULL : 0ULL;
        for (size_t w = startWord + 1; w < endWord; ++w) {
            data[w] = fillVal;
        }

        // 3. Modify partial end word
        uint64_t endMask = (endOffset == 63) ? ~0ULL : ((1ULL << (endOffset + 1)) - 1);
        if (value) {
            data[endWord] |= endMask;
        } else {
            data[endWord] &= ~endMask;
        }
    }
}

// Get a range of bits
std::vector<bool> BitArray::getRange(size_t min, size_t max) const {
    validate(min);
    validate(max);
    if (min > max) throw std::invalid_argument("min cannot be greater than max");
    
    std::vector<bool> result;
    result.reserve(max - min + 1);
    for (size_t i = min; i <= max; ++i) {
        result.push_back((data[i / 64] & (static_cast<uint64_t>(1) << (i % 64))) != 0);
    }
    return result;
}

// Find the first set bit (1)
size_t BitArray::findFirst() const {
    for (size_t i = 0; i < data.size(); ++i) {
        if (data[i] != 0) {
            return i * 64 + ctzll(data[i]);
        }
    }
    return static_cast<size_t>(-1); // Return -1 if no bit is set
}

// Count the number of set bits (1s)
size_t BitArray::countSetBits() const {
    size_t count = 0;
    for (size_t i = 0; i < data.size(); ++i) {
        count += popcntll(data[i]);
    }
    return count;
}

// Count the number of bytes used (non-zero bytes)
size_t BitArray::countBytesUsed() const {
    if (disposed) throw std::runtime_error("Storage is disposed");
    size_t count = 0;
    for (size_t i = 0; i < data.size(); ++i) {
        uint64_t word = data[i];
        // Branchless bitwise parallel check for non-zero bytes in a 64-bit word
        uint64_t temp = word | ((word & 0x7f7f7f7f7f7f7f7fULL) + 0x7f7f7f7f7f7f7f7fULL);
        count += popcntll(temp & 0x8080808080808080ULL);
    }
    return count;
}

// Serialize the BitArray to a JSON-like string
std::string BitArray::serialize() const {
    std::ostringstream oss;
    oss << "{\"bits\":" << bits << ",\"data\":[";
    for (size_t i = 0; i < data.size(); ++i) {
        oss << data[i];
        if (i < data.size() - 1) oss << ",";
    }
    oss << "]}";
    return oss.str();
}

// Deserialize a BitArray from a JSON-like string
BitArray BitArray::deserialize(const std::string& serialized) {
    size_t bitsPos = serialized.find("\"bits\":");
    if (bitsPos == std::string::npos) {
        throw std::invalid_argument("Malformed serialized data: missing \"bits\"");
    }
    size_t bitsStart = bitsPos + 7;
    size_t bitsEnd = serialized.find(",", bitsStart);
    if (bitsEnd == std::string::npos) {
        throw std::invalid_argument("Malformed serialized data: missing comma after bits");
    }
    size_t bits = std::stoull(serialized.substr(bitsStart, bitsEnd - bitsStart));

    size_t dataPos = serialized.find("\"data\":[");
    if (dataPos == std::string::npos) {
        throw std::invalid_argument("Malformed serialized data: missing \"data\"");
    }
    size_t dataStart = dataPos + 8;
    size_t dataEnd = serialized.find("]", dataStart);
    if (dataEnd == std::string::npos) {
        throw std::invalid_argument("Malformed serialized data: missing closing bracket for data");
    }
    std::string dataStr = serialized.substr(dataStart, dataEnd - dataStart);

    std::vector<uint64_t> dataVec;
    size_t pos = 0;
    if (!dataStr.empty()) {
        while (pos < dataStr.size()) {
            size_t commaPos = dataStr.find(",", pos);
            if (commaPos == std::string::npos) commaPos = dataStr.size();
            dataVec.push_back(static_cast<uint64_t>(std::stoull(dataStr.substr(pos, commaPos - pos))));
            pos = commaPos + 1;
        }
    }

    if (bits > std::numeric_limits<size_t>::max() - 63) {
        throw std::out_of_range("Deserialized bits size overflows capacity");
    }
    size_t expectedSize = bits / 64 + (bits % 64 != 0 ? 1 : 0);
    if (dataVec.size() != expectedSize) {
        throw std::invalid_argument("Malformed serialized data: data array size mismatch");
    }

    BitArray instance(bits);
    instance.data = std::move(dataVec);
    instance.clearUnusedBits();
    return instance;
}

// Get the total number of bits
size_t BitArray::getSize() const {
    return bits;
}

File: bitarray.js
// errors
const untitled_0 = new Error('The "size" parameter must be an integer and positive number');
const untitled_1 = new Error('Storage is disposed');
const untitled_2 = new Error('Size mismatch or invalid type');

const backend = require('./build/Release/bitarray.node');

const registry = new FinalizationRegistry(key => {
  try {
    backend.dispose(key);
  } catch {}
});

class BitArray {
  #disposed = false;
  #key;

  constructor(size, proxy = false) {
    if (!Number.isInteger(size) || size < 0 || size > 0xFFFFFFFF) throw untitled_0;
    this.#key = backend.create(size);
    registry.register(this, this.#key, this);

    if (proxy) {
      const isIndex = (p) => {
        if (typeof p !== 'string') return false;
        const int = Number(p);
        return Number.isInteger(int) && String(int) === p && int >= 0;
      };

      return new Proxy(this, {
        get: (target, p) => {
          if (isIndex(p)) {
            return backend.get(this.#key, Number(p));
          }
          const val = Reflect.get(target, p);
          if (typeof val === 'function') {
            return val.bind(target);
          }
          return val;
        },
        set: (target, p, v) => {
          if (isIndex(p)) {
            backend.set(this.#key, Number(p), v);
            return true;
          }
          if (p === 'length') {
            target.resize(v);
            return true;
          }
          return Reflect.set(target, p, v);
        }
      });
    }
  }

  #isDisposed() { if (this.#disposed) throw untitled_1; }

  set(pos, bool) {
    this.#isDisposed();
    backend.set(this.#key, pos, bool);
  }

  get(pos) {
    this.#isDisposed();
    return backend.get(this.#key, pos);
  }

  resize(size) {
    if (!Number.isInteger(size) || size < 0 || size > 0xFFFFFFFF) throw untitled_0;
    this.#isDisposed();
    backend.resize(this.#key, size);
  }

  bitwiseOp(other, op) {
    if (!(other instanceof BitArray) || this.size !== other.size) throw untitled_2;
    this.#isDisposed();
    backend[op](this.#key, other.#key);
  }

  bitwiseAnd(other) {
    this.#isDisposed();
    this.bitwiseOp(other, 'bitwiseAnd');
  }
  bitwiseOr(other) {
    this.#isDisposed();
    this.bitwiseOp(other, 'bitwiseOr');
  }
  bitwiseXor(other) {
    this.#isDisposed();
    this.bitwiseOp(other, 'bitwiseXor');
  }

  bitwiseNot() {
    this.#isDisposed();
    backend.bitwiseNot(this.#key);
  }

  setBatch(positions, bool) {
    this.#isDisposed();
    backend.setBatch(this.#key, positions, bool);
  }

  getBatch(positions) {
    this.#isDisposed();
    return backend.getBatch(this.#key, positions);
  }

  setRange(min, max, bool) {
    this.#isDisposed();
    backend.setRange(this.#key, min, max, bool);
  }
  getRange(min, max) {
    this.#isDisposed();
    return backend.getRange(this.#key, min, max);
  }

  findFirst() {
    this.#isDisposed();
    return backend.findFirstSet(this.#key);
  }

  countSetBits() {
    this.#isDisposed();
    return backend.countSetBits(this.#key);
  }

  countBytesUsed() {
    this.#isDisposed();
    return backend.countBytesUsed(this.#key);
  }

  serialize() {
    this.#isDisposed();
    return backend.serialize(this.#key);
  }

  static deserialize(serialized) {
    const instance = new BitArray(1);
    try {
      const key = backend.deserialize(serialized);
      registry.unregister(instance);
      backend.dispose(instance.#key);
      instance.#key = key;
      registry.register(instance, key, instance);
      return instance;
    } catch (err) {
      instance.dispose();
      throw err;
    }
  }
  static from(serialized) { return BitArray.deserialize(serialized); }

  dispose() {
    if (this.#disposed) return;
    this.#disposed = true;
    registry.unregister(this);
    try {
      backend.dispose(this.#key);
    } catch {}
  }

  get size() { return backend.getSize(this.#key); }
  get length() { return backend.getSize(this.#key); }

  *[Symbol.iterator]() {
    const length = this.size;
    const chunkSize = 1024;
    for (let i = 0; i < length; i += chunkSize) {
      const end = Math.min(i + chunkSize - 1, length - 1);
      const chunk = backend.getRange(this.#key, i, end);
      for (let j = 0; j < chunk.length; j++) {
        yield chunk[j];
      }
    }
  }
}

module.exports = BitArray;
